http-driver/src/main/java/org/neo4j/ogm/drivers/http/response/AbstractHttpResponse.java

```diff
@@ -32,81 +35,152 @@
+        try (JsonParser jsonParser = JSON_FACTORY.createParser(inputStream)) {
+
+            // First start parsing
+            jsonParser.nextToken();
+            // Then copy the whole thing
+            return createTokenBuffer(jsonParser);
+        }
+    }
+
+    private TokenBuffer createTokenBuffer(JsonParser jsonParser) throws IOException {
+
+        TokenBuffer tokenBuffer = new TokenBuffer(mapper, true);
+        tokenBuffer.copyCurrentStructure(jsonParser);
+
+        return tokenBuffer;
+    }
+
+    private static JsonParser findNextObject(JsonParser parser, String nodeName) throws IOException {
+
+        JsonToken jsonToken;
+        while ((jsonToken = parser.nextToken()) != null) {
+            parser.skipChildren();
+            if (JsonToken.FIELD_NAME.equals(jsonToken) && parser.getCurrentName().equals(nodeName)) {
+                parser.nextToken();
+                return parser;
             }
-        } catch (IOException e) {
-            throw new ResultProcessingException("Error processing results", e);
         }
+        return null;
+    }
+
+    private void throwExceptionOnErrorEntry(JsonParser pointingToErrors) throws IOException {
+
+        if (pointingToErrors == null) {
+            return;
+        }
+
+        JsonNode errorsNode = mapper.readTree(pointingToErrors);
+        Optional<JsonNode> optionalErrorNode = StreamSupport.stream(errorsNode.spliterator(), false)
+            .findFirst();
+        if (optionalErrorNode.isPresent()) {
+            JsonNode errorNode = optionalErrorNode.get();
+            throw new CypherException(errorNode.findValue("code").asText(), errorNode.findValue("message").asText());
+        }
+    }
+
+    private String[] readColumns(TokenBuffer bufferedResults) throws IOException {
+
+        JsonParser parser = bufferedResults.asParserOnFirstToken();
+        parser.nextToken();
+        parser = findNextObject(parser, "columns");
+        if (parser == null) {
+            return new String[0];
+        } else {
+            return mapper.readValue(parser, String[].class);
+        }
+    }
+
+    private QueryStatistics readQueryStatistics(TokenBuffer bufferedResults) throws IOException {
+
+        JsonParser parser = bufferedResults.asParserOnFirstToken();
+        parser.nextToken();
+        parser = findNextObject(parser, "stats");
+        if (parser != null) {
+            return mapper.readValue(parser, QueryStatisticsModel.class);
+        }
+
+        return null;
     }
 
     public T nextDataRecord(String key) {
-        JsonToken token;
         try {
-            while ((token = bufferParser.nextToken()) != null) {
-                if (JsonToken.FIELD_NAME.equals(token)) {
-                    if (key.equals(bufferParser.getCurrentName())) {
-                        return mapper.readValue(bufferParser, resultClass);
-                    }
-                }
+            while (results.hasNext()) {
```

Comment:
This loop is not needed but could be an if-statement, or? The `t` gets always returned if there is a next.